crates/storage-query-datafusion/src/partition_store_scanner.rs

```diff
@@ -36,14 +37,31 @@ pub trait ScanLocalPartition: Send + Sync + Debug + 'static {
     fn append_row(row_builder: &mut Self::Builder, string_buffer: &mut String, value: Self::Item);
 }
 
+// ScanLocalPartitionInPlace can be used for large or performance sensitive tables to build batches on io threads (instead of sending rows cross-thread)
+pub trait ScanLocalPartitionInPlace: Send + Sync + Debug + 'static {
+    type Builder: crate::table_util::Builder + Send;
+    type Item: Send;
+
+    fn for_each_row<F: FnMut(Self::Item) -> ControlFlow<()> + Send + Sync + 'static>(
+        partition_store: &PartitionStore,
+        range: RangeInclusive<PartitionKey>,
+        f: F,
+    ) -> Result<impl Future<Output = restate_storage_api::Result<()>> + Send, StorageError>;
+
+    fn append_row(row_builder: &mut Self::Builder, string_buffer: &mut String, value: Self::Item);
+}
+
+pub struct ScanStream;
```

Comment:
this is a poor attempt at specialisation with marker types. what we really want is to allow there to be two scan_partition impls (one in place and one streaming), and have preference for the in place one. as this is not really possible, i instead make it so that we declare at scanner creation time that we want to use the in place one. this seems to work ok.

Reply:
What would be the main benefit from having the streaming version?